src/common/services/session/session.service.js

```diff
@@ -46,95 +76,297 @@ const session = /* @ngInject */ function ($cookies, $rootScope, $http, $timeout,
   return {
     session: session,
     sessionSubject: sessionSubject,
+    authClient: authClient, // Exposed for tests only
+    oktaSignInWidgetDefaultOptions,
+    catchCreateAccountErrors: catchCreateAccountErrors,
+    clearCheckoutSavedData: clearCheckoutSavedData,
+    checkCreateAccountStatus: checkCreateAccountStatus,
+    createAccount: createAccount,
+    downgradeToGuest: downgradeToGuest,
     getRole: currentRole,
+    getOktaUrl: getOktaUrl,
+    handleOktaRedirect: handleOktaRedirect,
+    hasLocationOnLogin: hasLocationOnLogin,
+    isOktaRedirecting: isOktaRedirecting,
+    oktaIsUserAuthenticated: oktaIsUserAuthenticated,
+    updateCurrentProfile: updateCurrentProfile,
+    updateCheckoutSavedData: updateCheckoutSavedData,
+    removeOktaRedirectIndicator: removeOktaRedirectIndicator,
+    removeLocationOnLogin: removeLocationOnLogin,
     signIn: signIn,
     signOut: signOut,
-    signUp: signUp,
-    forgotPassword: forgotPassword,
-    resetPassword: resetPassword,
-    downgradeToGuest: downgradeToGuest
+    signOutWithoutRedirectToOkta: signOutWithoutRedirectToOkta
   }
 
-  /* Public Methods */
-  function signIn (username, password, mfa_token, trust_device, lastPurchaseId) {
-    const data = {
-      username: username,
-      password: password
-    }
-    if (angular.isDefined(mfa_token)) { data.mfa_token = mfa_token }
-    if (trust_device) { data.trust_device = '1' }
-    // Only send lastPurchaseId if present and currently public
-    if (angular.isDefined(lastPurchaseId) && currentRole() === Roles.public) { data.lastPurchaseId = lastPurchaseId }
-    return Observable
-      .from($http({
-        method: 'POST',
-        url: casApiUrl('/login'),
-        data: data,
-        withCredentials: true
+  function handleOktaRedirect (lastPurchaseId) {
+    if (authClient.isLoginRedirect()) {
+      return Observable.from(authClient.token.parseFromUrl().then((tokenResponse) => {
+        authClient.tokenManager.setTokens(tokenResponse.tokens)
+        return signIn(lastPurchaseId)
       }))
-      .map((response) => response.data)
+    } else {
+      return Observable.of(false)
+    }
+  }
+
+  function signIn (lastPurchaseId) {
+    session.isOktaRedirecting = true
+    return Observable.from(internalSignIn(lastPurchaseId))
+      .map((response) => response ? response.data : response)
       .finally(() => {
         $rootScope.$broadcast(SignInEvent)
       })
   }
 
-  function signOut () {
-    // https://github.com/CruGlobal/cortex_gateway/wiki/Logout
-    return Observable
-      .from($http({
-        method: 'DELETE',
-        url: casApiUrl('/logout'),
+  async function internalSignIn (lastPurchaseId) {
```

Comment:
A lot of the new code in this file uses async/await. Native JS promises don't trigger a AngularJS digest cycle in their `then`/`catch` handlers, so we should use observables instead.

Reply:
I have recreated the functions to use observables instead of promises.

Thank you for spotting this; it allowed me to improve the code.